contracts/ResilientOracle.sol

```diff
@@ -200,35 +199,82 @@ contract ResilientOracle is OwnableUpgradeable, PausableUpgradeable, ResilientOr
-        // Global emergency switch
-        require(!paused(), "resilient oracle is paused");
+    function _getMainOraclePrice(address vToken) internal view returns (uint256) {
+        uint256 price = INVALID_PRICE;
 
         (address mainOracle, bool mainOracleEnabled) = getOracle(vToken, OracleRole.MAIN);
+        if (mainOracleEnabled && mainOracle != address(0)) {
+            try OracleInterface(mainOracle).getUnderlyingPrice(vToken) returns (uint256 mainOraclePrice) {
+                price = mainOraclePrice;
 
-        uint256 price = INVALID_PRICE;
+                (address pivotOracle, bool pivotOracleEnabled) = getOracle(vToken, OracleRole.PIVOT);
 
-        if (!mainOracleEnabled) {
-            return price;
+                if (pivotOracleEnabled && pivotOracle != address(0)) {
+                    try OracleInterface(pivotOracle).getUnderlyingPrice(vToken) returns (uint256 pivotPrice) {
+                        if (pivotPrice != INVALID_PRICE) {
+                            bool isPriceValid = boundValidator.validatePriceWithAnchorPrice(vToken, price, pivotPrice);
+                            if (!isPriceValid) {
+                                return INVALID_PRICE;
+                            }
+                        }
+                    } catch {}
+                }
+            } catch {}
         }
 
-        try OracleInterface(mainOracle).getUnderlyingPrice(vToken) returns (uint256 _price) {
-            price = _price;
+        return price;
+    }
+
+    /**
+     * @notice This function won't revert when price is 0, because the getUnderlyingPrice checks if pirce is > 0
+     * @param vToken vToken address
+     * @return price USD price in 18 decimals
+     */
+    function _getFallbackOraclePrice(address vToken) internal view returns (uint256) {
+        uint256 price = INVALID_PRICE;
+        bool compareWithMain = false;
 
-            (address pivotOracle, bool pivotOracleEnabled) = getOracle(vToken, OracleRole.PIVOT);
+        (address fallbackOracle, bool fallbackEnabled) = getOracle(vToken, OracleRole.FALLBACK);
+        if (fallbackEnabled && fallbackOracle != address(0)) {
+            try OracleInterface(fallbackOracle).getUnderlyingPrice(vToken) returns (uint256 fallbackOraclePrice) {
+                price = fallbackOraclePrice;
 
-            // Price oracle is not mandantory
-            if (pivotOracle == address(0) || !pivotOracleEnabled) {
-                return price;
-            }
+                (address pivotOracle, bool pivotOracleEnabled) = getOracle(vToken, OracleRole.PIVOT);
+                if (pivotOracleEnabled && pivotOracle != address(0)) {
+                    try OracleInterface(pivotOracle).getUnderlyingPrice(vToken) returns (uint256 pivotPrice) {
+                        if (pivotPrice != INVALID_PRICE) {
+                            bool isPriceValid = boundValidator.validatePriceWithAnchorPrice(vToken, price, pivotPrice);
```

Comment:
can we clarify this logic? So if the priceIs not valid ( comparing main and pivot), what do we do then?

Reply:
If main and pivot validation fails then we compare fallback and main prices. If fallback and main prices are validated then we return the main price.